File: Java/AlgoExpert/Dynamic Programming/Medium/Levenshtein Distance.java
import java.util.*;

class Program {
  public static int levenshteinDistance(String str1, String str2) {
    // Write your code here.
		int[][] grid = new int[str1.length() + 1][str2.length() + 1];
		
		for (int column = 0; column < grid[0].length; column++) {
			grid[0][column] = column;
		}
		
		for (int row = 0; row < grid.length; row++) {
			grid[row][0] = row;
		}
		
		for (int i = 1; i < grid.length; i++) {
			for (int j = 1; j < grid[i].length; j++) {
				char str1Char = str1.charAt(i - 1);
				char str2Char = str2.charAt(j - 1);
				
				if (str1Char == str2Char) {
					grid[i][j] = grid[i - 1][j - 1];
				} else {
					grid[i][j] = 1 + Math.min(grid[i - 1][j - 1], Math.min(grid[i - 1][j], grid[i][j - 1]));
				}
			}
		}
		
    return grid[str1.length()][str2.length()];
  }
}

File: Python/AlgoExpert/Dynamic Programming/Medium/levenshteinDistance.py
def levenshteinDistance(str1, str2):
    # Write your code here.
    grid = [[x for x in range(len(str1) + 1)] for y in range(len(str2) + 1)]
    
    for i in range(len(str2) + 1):
        grid[i][0] = i
    
    for row in range(1, len(str2) + 1):
        for col in range(1, len(str1) + 1):
            if str2[row - 1] == str1[col - 1]:
                grid[row][col] = grid[row - 1][col - 1]
            else:
                grid[row][col] = min(grid[row - 1][col], grid[row][col - 1], grid[row - 1][col - 1]) + 1

    return grid[-1][-1]
